src/ao_rocksdb_store.erl

```diff
@@ -0,0 +1,493 @@
+-spec resolve(Opts, Path) -> Result
+    when Opts :: any(),
+         Path :: binary() | list(),
+         Result :: not_found | binary() | [binary() | list()].
+resolve(Opts, [BaseId, ResolutionKeys]) ->
+    case resolve_chain(Opts, BaseId, ResolutionKeys) of
+        not_found ->
+            fallback_resolve(Opts, BaseId, ResolutionKeys);
+        Result ->
+            maybe_remove_messages_prefix(Result)
+    end;
+resolve(_Opts, Key) when is_binary(Key) ->
+    case item_path(Key) of
+        not_found ->
+            not_found;
+        {ok, [LastKey | _]} ->
+            maybe_remove_messages_prefix(LastKey)
+    end.
+
+item_path(Key) ->
+    gen_server:call(?MODULE, {item_path, Key}, ?TIMEOUT).
+
+%% Helper function to resolve a chain of IDs.
+resolve_chain(Opts, BaseId, ResolutionKeys) ->
+    lists:foldl(fun (_Key, not_found) ->
+                        not_found;
+                    (Key, Acc) ->
+                        case read(Opts, Acc) of
+                            not_found ->
+                                not_found;
+                            {ok, ItemBinary} ->
+                                extract_next_id(ItemBinary, Key)
+                        end
+                end,
+                BaseId,
+                ResolutionKeys).
+
+%% Helper function to resolve the next ID from an item.
+extract_next_id(ItemBinary, ResolutionKey) ->
+    Item = ar_bundles:deserialize(ItemBinary),
+    Manifest =
+        ar_bundles:parse_manifest(
+            maps:get(<<"manifest">>, Item#tx.data)),
+    NextId = maps:get(list_to_binary(ResolutionKey), Manifest),
+    <<"messages-", NextId/binary>>.
+
+%% Fallback resolution if primary resolution fails.
+fallback_resolve(Opts, BaseId, ResolutionKeys) ->
+    Path = path(Opts, ResolutionKeys),
+    case item_path(<<BaseId/binary, "-", Path/binary>>) of
+        not_found ->
+            not_found;
+        {ok, [Key | _]} ->
+            maybe_remove_messages_prefix(Key)
+    end.
+
+-spec type(Opts, Key) -> Result
+    when Opts :: map(),
+         Key :: binary(),
+         Result :: composite | simple | not_found.
+type(Opts, ["messages", Path]) ->
+    MessagesPath = <<"messages-", Path/binary>>,
+    type(Opts, MessagesPath);
+type(Opts, Key) when is_binary(Key) ->
+    % The fs adapter looks on the FS. If it's a directory it says 'comosite'
+    % composite item is an item which has manifest!
+    case read(Opts, Key) of
+        not_found ->
+            not_found;
+        {ok, ItemBinary} ->
+            Item = ar_bundles:deserialize(ItemBinary),
+            case ar_bundles:manifest(Item) of
+                undefined ->
+                    simple;
+                _Manifest ->
+                    composite
+            end
+    end.
+
+make_group(_Opts, _Path) ->
+    % FS just creates a folder....
+    % when we write a composite item
+    % (the item which consists of multiple items) we do:
+    % 1. create folder
```

Comment:
I am not sure I follow what this comment is trying to say. Is it replicating the comment in `write_composite` in ao_cache? Having the function stubbed out making sense though.

Reply:
Sorry, I think it's a leftover that should be removed.